### PR TITLE
Feature/support dfs

### PR DESCRIPTION
**Pull request by iddohau**
Add support for dfs.

### CODE REVIEW

File: smbprotocol/structure.py [hunk 1]
```diff
@@ -932,3 +932,23 @@ def _get_packed_size(self):
 
     def _to_string(self):
         return self.value
+
+
+class NullTerminatedTextField(TextField):
+
+    def set_value(self, value):
```
[the remark this answers is not in the document]
Comment (iddohau) [reply smbprotocol/structure.py:939]: I will change this.

**Comment by iddohau** [comment]
> I'm curious as to how this all works as I'm not too well versed in DFS and how it all works. Currently it seems like in the smbclient land you are just automatically redirecting a share that is a DFS share to the proper hostname but the user would still need to specify an actual host in the UNC path?

The logic is:
If the server is DFS I will query the DFS server with referral request.
The reponse will give me the network path of the SMB server.
Then I will connect to the SMB server instead of the DFS server.

**Comment by jborean93** [comment]
Really do apologize for the delay here, I've started to look into this this week and hope to have this in before the 1.1.0 release. From a brief overview this looks like it will support DFS if you are connecting to an actual DFS root in the path but doesn't support something like `\\domain.com\dfs_path` like you can on Windows. It sounds like that will require further work to lookup the `ldap` SRV record of that domain, somehow query AD to get a list of DFS root servers then connect to those for the referral request. I'm playing around with this a bit more to see if I can support both scenarios but will settle with what's here for now if that's too hard to implement.

**Comment by jborean93** [comment]
Thanks @iddohau for starting the work on this one. I created a new PR based on your initial changes here and have just merged it into master https://github.com/jborean93/smbprotocol/pull/49.

There were a few more changes that needed to be done to accommodate domain based DFS shares that I came across in my testing and I found it prudent to implement a rudimentary form of caching of referrals as well. DFS support will be present in the 1.2.0 release at some point soon.

**Pull request closed without merging**